## Урок №3. Ввод-вывод и базовые переменные

**Задание №1**

Давай представим, что мы пишем бэкенд для сайта ветеринарной клиники. Нам нужно написать программу, которая будет запрашивать у пользователя вид питомца, его возраст и кличку, а потом выведет все эти данные в одно предложение. Например:

`Это желторотый питон по кличке "Каа". Возраст: 34 года.`

**Решение задания № 1:**

In [1]:
KindOfPet=str(input('Вид питомца? '))
NameOfPet=str(input('Кличка питомца? '))
NameOfPet='"'+NameOfPet+'"'
Age=str(input('Возраст питомца? '))
print ('Это',KindOfPet,'по кличке',NameOfPet,'. Возраст:',Age )

Вид питомца? желторотвй питон
Кличка питомца? Каа
Возраст питомца? 34 года
Это желторотвй питон по кличке "Каа" . Возраст: 34 года


**Задание №2**

А теперь мы с тобой напишем форму ввода ответа на тест по биологии для студентов. Он должен запрашивать по порядку этапы развития человека (проверим твое умение гуглить, что тоже очень важно для программиста. ) и в конце вывести все стадии, разделенные знаком `=>`, что будет означать постепенный переход от одного к другому. В следующих уроках мы дополним эту форму до полноценного теста, который будет проверять правильность ответов, а пока - начнем с малого. Напоминаем, что разделить эти данные тебе поможет команда `sep` внутри команды `print`, например, чтобы разделить переменные знаком `+` нужно ввести: 

`print(a1, a2, a3, sep='+')`

Подсказка: последняя стадия развития - Homo sapiens sapiens.

**Решение задания № 2:**

In [2]:
a1=input('1-й этап?')#Australopithecus
a2=input('2-й этап?')#Homo habilis
a3=input('3-й этап?')#Homo erectus
a4=input('4-й этап?')#Homo neanderthalensis
a5=input('5-й этап?')#Homo sapiens sapiens
print(a1,a2,a3,a4,a5,sep=' => ')

1-й этап?Australopithecus
2-й этап?Homo habilis
3-й этап?Homo erectus
4-й этап?Homo neanderthalensis
5-й этап?Homo sapiens sapiens
Australopithecus => Homo habilis => Homo erectus => Homo neanderthalensis => Homo sapiens sapiens


## Урок №4. float, int и арифметические операции

**Задание №1**

Пользователь вводит стороны прямоугольника, выведите его площадь и периметр. На вход программе могут подаваться как целые числа, так и вещественные

**Решение задания № 1:**

In [4]:
DL=float(input('Длина прямоугольника?'))
SH=float(input('Ширина прямоугольника?'))
print('Площадь прямоугольника =', DL*SH, 'Периметр прямоугольника =',2*(DL+SH))

Длина прямоугольника?4.36
Ширина прямоугольника?2
Площадь прямоугольника = 8.72 Периметр прямоугольника = 12.72


**Задание №2**

Дано пятизначное целое число. Напишите алгоритм, который возведёт количество десятков в степень количества единиц. Затем умножит это число на количество сотен. И делит получившееся число на разность количества десятков тысяч и количества тысяч

Например, есть число `46275`

Необходимо возвести 7 (десятки) в степень 5 (единицы), умножить получившееся число на 2 (сотни), и разделить на разность между 4 (десятки тысяч) и 6 (тысячи) то есть (4-6)

В результате необходимо получить вещественное число. В нашем примере это будет: `-16807.0`

**Решение задания № 2:**

In [5]:
AllNumber=int(input('Введите пятизначное число...'))
s=str(AllNumber)
l=len(list(s))
ed=int((s[(l-1)]))
des=int((s[(l-2)]))
sot=int((s[(l-3)]))
tys=int((s[(l-4)]))
destys=int((s[(l-5)]))
OutNumber=des**ed
OutNumber=OutNumber*sot
OutNumber=OutNumber/(destys-tys)
print(OutNumber)

Введите пятизначное число...46275
-16807.0


## Урок №5. Логические и условные операторы

**Задание №1**

Пользователь вводит целое число. Выведите его строку-описание вида `"отрицательное четное число"`, `"нулевое число"`, `"положительное нечетное число"`, например, численным описанием числа 190 является строка `"положительное четное число"`. Если число не является четным - выведите сообщение `"число не является четным"`

**Решение задания № 1:**

In [6]:
UsersNumber=int(input('Введите целое число:'))
MyOutString=' число'
if (UsersNumber/2==UsersNumber//2):
    MyOutString=' четное'+MyOutString
else:
    MyOutString=' нечетное'+MyOutString
if (UsersNumber==0):
    MyOutString='нулевое число'
elif(UsersNumber<0):
    MyOutString= 'отрицательное'+MyOutString
else:
    MyOutString= 'положительное'+MyOutString
print(MyOutString)
if (UsersNumber/2!=UsersNumber//2):
    print('число не является четным')


Введите целое число:190
положительное четное число


**Задание №2**

Дано слово из маленьких латинских букв. Сколько там согласных и гласных букв? Гласными называют буквы `«a», «e», «i», «o», «u»`.

Для решения задачи создайте переменную и в неё положите слово с помощью `input()`

А также определите количество каждой из этих гласных букв
Если какой-то из перечисленных букв нет - Выведите False

**Решение задания № 2:**

In [7]:
ia,ie,ii,io,iu,ic=0,0,0,0,0,0
MyEnlishWord=input('Input any word on  an English keyboard in low case, please:')
for lit in MyEnlishWord:
    if(lit=='a'):
        ia+=1
    elif(lit=='e'):
        ie+=1
    elif(lit=='i'):
        ii+=1
    elif(lit=='o'):
        io+=1
    elif(lit=='u'):
        iu+=1
    else:
        ic=ic+1
NumVow=ia+ie+ii+io+iu
print('there are',str(NumVow),' vowel letters in your word')
print('there are',str(ic),' consonant letters in your word')
print('About vowel letters:')
if ia==0:
    print('a- False')
else:
    print('a-',str(ia))
if ie==0:
    print('e- False')
else:
    print('e-',str(ie))   
if ii==0:
    print('i- False')
else:
    print('i-',str(ii))
if io==0:
    print('o- False')
else:
    print('o-',str(io))
if iu==0:
    print('u- False')
else:
    print('u-',str(iu))

Input any word on  an English keyboard in low case, please:looking
there are 3  vowel letters in your word
there are 4  consonant letters in your word
About vowel letters:
a- False
e- False
i- 1
o- 2
u- False


**Задание №3**

Два инвестора - Майкл и Иван хотят вложиться в стартап. Фаундеры сказали, что минимальная сумма инвестиций - `X` долларов, больше инвестировать можно сколько угодно. У Майкла `A` долларов, у Ивана `B` долларов. Если оба могут вложиться - выведите `2`, если только Майкл - `Mike`, если только Иван - `Ivan`, если не могут по отдельности, но вместе им хватает - `1`, если никто - `0`.

**Решение задания № 3:**

In [10]:
X=int(input('Минимальная сумма инвестиций мтартапа?:'))
A=int(input('Сколько долларов у Майкла ?:'))
B=int(input('Сколько долларов у Ивана ?:'))
if (A>=X and B>=X):
    print(2)
elif (A>=X and B<X):
    print('Mike')
elif (A<X and B>=X):
    print('Ivan')
elif ((A+B)>=X):
    print(1)
else:
    print(0)


Минимальная сумма инвестиций мтартапа?:77
Сколько долларов у Майкла ?:120
Сколько долларов у Ивана ?:50
Mike


## Урок №6. Циклы while и for

**Задание №1**

Cначала вводится число N, затем вводится ровно N целых чисел. Подсчитайте, сколько из них равны нулю, и выведите это количество.

**Решение задания № 1:**

In [ ]:
CountNull=0
N=int(input('Сколько целых чисел будем вводить?:'))
for i in range(1,N+1,1):
    CurrentNumber=int(input('Введите целое число №'+str(i)+": "))
    if CurrentNumber==0:
        CountNull=CountNull+1
print('Колличество введенных нулевых чисел равно ',str(CountNull))

**Задание №2**

Вводится натуральное число X. Подсчитайте количество натуральных делителей числа X (включая 1 и само число). x ≤ 2e9 (2 миллиарда)

**Решение задания № 2:**

In [11]:
X=int(input('Введите натуральное число:'))
OutNumberOfTask=0
for i in range(1,X+1,1):
    if(X/i==X//i):
        OutNumberOfTask=OutNumberOfTask+1
print('Количество натуральных делителей этого числа рвно ',OutNumberOfTask)

Введите натуральное число:4
Количество натуральных делителей этого числа рвно  3


**Задание №3**

Вводятся целые числа A и B. Гарантируется, что A ≤ B. Выведите все четные числа на заданном отрезке через пробел.

**Решение задания № 3:**

In [13]:
A=int(input('Введите начальное целое число диапазона для обработки:'))
B=int(input('Введите конечное целое число диапазона для обработки:'))
OutString=""
for i in range(A,B,1):
    if (i/2==i//2):
        OutString=OutString+' '+str(i)
print('Четные числа в заданном диапазоне -',OutString)

Введите начальное целое число диапазона для обработки:1
Введите конечное целое число диапазона для обработки:10
Четные числа в заданном диапазоне -  2 4 6 8


## Урок №7. Строки

**Задание №1**

На вход подается 1 строка без пробелов. По данной строке определите, является ли она палиндромом (то есть, можно ли прочесть ее наоборот, как, например, слово "шалаш"). Необходимо вывести ”yes”, если строка является палиндромом, и “no” в противном случае.

**Решение задания № 1:**

In [18]:
InString=input('Введите исходную строку: ')
if InString==InString[::-1]:
    print('yes')
else:
    print('no')

Введите исходную строку: шалаш
yes


**Задание №2**

Дана строка, длина которой не превосходит 1000. Вам требуется преобразовать все идущие подряд пробелы в один. Выведите измененную строку.

**Решение задания № 2:**

In [47]:
InString=input('Введите исходную строку: ')
#InString="Гори       горри   моя звезда. Hello my       sweet world!"
InString="  "+InString
DlStr=len(InString)
OutString=""
Pr=" "
for i in range(DlStr-1,1,-1):
    CurrSymb=InString[i]
    NextSymb=InString[i-1]
    if (CurrSymb==Pr and NextSymb==Pr):
        p=7
    else:
        OutString=CurrSymb+OutString    
print(OutString)

Введите исходную строку: Гори       горри   моя звезда. Hello my       sweet world!
Гори горри моя звезда. Hello my sweet world!


## Урок №8. Списки

**Задание №1**

В первой строке вводится число N. Далее в N строк вводится N чисел (1 ≤ N ≤ 10000), по одному числу на строке. Все числа по модулю не превышают 10e5. Переверните массив чисел. Выведите N чисел - перевернутый массив.

**Решение задания № 1:**

In [78]:
N=int(input('Введите количество элементов в списке:'))
MyList=[]
for i in range(0,N,1):
    MyList.append(input('Введите элемент №'+str(i+1)+' '))
MyList.reverse ()   
print(MyList)    

Введите количество элементов в списке:3
Введите элемент №1 11
Введите элемент №2 7
Введите элемент №3 3
['3', '7', '11']


**Задание №2**

В первую строчку вводится число N (1 ≤ N ≤ 100 000). В следующую строку через пробел вводятся N чисел (1 ≤ Ai ≤ 10e9). Вам требуется написать метод, который получает на вход массив и изменяет его таким образом, чтобы на первом месте стоял последний элемент, на втором - первый, на третьем - второй и т. д. Выведите N чисел - измененный массив. 

**Решение задания № 2:**

In [7]:
def MyNewList(MyList):
    First=MyList[0]
    MyList.pop(0)
    MyList.append(First)
    return MyList
N=int(input('Введите количество элементов в списке:'))
MyList=input('Введите  элементы в одну строку серез пробелы:').split()
print(MyList)
MNL=MyNewList(MyList)
print(MNL)

Введите количество элементов в списке:3
Введите  элементы в одну строку серез пробелы:45 456 8
['45', '456', '8']
['456', '8', '45']


**Задание №3**

На берегу реки стояли n рыбаков, все они хотели перебраться на другой берег. Одна лодка может выдержать не более m килограмм, при этом в лодку помещается не более 2 человек. Определите, какое минимальное число лодок нужно, чтобы перевезти на другой берег всех рыбаков
В первую строку вводится число m (1 ≤ m ≤ 10e6) - максимальная масса, которую может выдержать одна лодка. Во вторую строку вводится число n (1 ≤ n ≤ 100) - количество рыбаков. В следующие N строк вводится по одному числу Ai (1 ≤  Ai ≤ m) - вес каждого путешественника.
Программа должна вывести одно число - минимальное количество лодок, необходимое для переправки всех рыбаков на противоположный берег.

**Решение задания № 3:**

In [26]:
m=int(input('Введите максимальную грузоподъемность одной лодки в килограммах:'))
n=int(input('Введите количество рыбаков:'))
ListWeights=input('Введите  в одну строку серез пробелы вес каждого путешественника:').split()
t=sorted(ListWeights,reverse=True)
c=0
if int(t[0]) > m:
    print('Задача не имеет решения')
    exit()
while len(t):
    c += 1
    k = m - int(t.pop(0))
    for i in range(len(t)):
        if int(t[i]) <= k:
            t.pop(i)
            break
print('Минимальное количество лодок равно -',str(c))

Введите максимальную грузоподъемность одной лодки в килограммах:100
Введите количество рыбаков:3
Введите  в одну строку серез пробелы вес каждого путешественника:78 87 65
Минимальное количество лодок равно - 3


В цикле, пока рыбаки не кончились:

1. Увеличиваем кол-во лодок на 1.
2. Сажаем в лодку самого тяжёлого из оставшихся рыбаков.
3. Среди оставшихся рыбаков ищем самого тяжёлого, вес которого не превышает оставшуюся после посадки первого рыбака грузоподъёмность лодки. Если нашли - сажаем его на второе место в лодке.

## Урок №9. Множества

**Задание №1**

В первую строку вводится число N – количество чисел (1 ≤ N ≤ 100000). Во вторую строку вводится через пробел N чисел, каждое не превышает 2*10e9 по модулю. Требуется выяснить, сколько среди этих чисел различных. Выведите число, равное количеству различных чисел среди данных.

**Решение задания № 1:**

In [2]:
N=int(input('Введите количество вводимых чисел:'))
ListOfNumbers=input('Введите  в одну строку серез пробелы числа:').split()
MySet=set(ListOfNumbers)
print('Количество различных чисел среди введенных равно ',len(MySet))

Введите количество вводимых чисел:7
Введите  в одну строку серез пробелы числа:11 0 10 11 0 0 11 
Количество различных чисел среди введенных равно  3


**Задание №2**

Вводятся два списка чисел, которые могут содержать до 100000 чисел каждый. Все числа каждого списка находятся на отдельной строке. Выведите, сколько чисел содержится одновременно как в первом списке, так и во втором.

**Решение задания № 2:**

In [29]:
def create_my_lists(num_list):
    counter_list,str_input =1,"1"
    my_list=list()
    while ((counter_list< 10) and(str_input!="")):
        str_input=input('Введите число номер ' +str(counter_list)+' списка номер '+str(num_list)+', для окончания ввода списка введите Enter ')
        counter_list=counter_list+1
        my_list.append(str_input)
    my_list.pop()
    return my_list

My_list_one=create_my_lists(1)
My_list_two=create_my_lists(2)
My_set_one=set(My_list_one)
My_set_two=set(My_list_two)
My_crossing_set=My_set_one & My_set_two
print('Количество чисел, которые содержутся как в первом списке, так и во втором равно ',str(len(My_crossing_set)))  

Введите число номер 1 списка номер 1, для окончания ввода списка введите Enter 45
Введите число номер 2 списка номер 1, для окончания ввода списка введите Enter 78
Введите число номер 3 списка номер 1, для окончания ввода списка введите Enter 4
Введите число номер 4 списка номер 1, для окончания ввода списка введите Enter 
Введите число номер 1 списка номер 2, для окончания ввода списка введите Enter 78
Введите число номер 2 списка номер 2, для окончания ввода списка введите Enter 2
Введите число номер 3 списка номер 2, для окончания ввода списка введите Enter 25
Введите число номер 4 списка номер 2, для окончания ввода списка введите Enter 
Количество чисел, которые содержутся как в первом списке, так и во втором равно  1


**Задание №3**

Во входную строку водится последовательность чисел через пробел. Для каждого числа выведите слово ”YES” (в отдельной строке), если это число ранее встречалось в последовательности или ”NO”, если не встречалось.

**Решение задания № 3:**

In [31]:
ListOfNumbers=input('Введите  в одну строку серез пробелы числа:').split()
MyList=list()
MySet=set()
NumberOfDoublesFound=0
for MyNumber in ListOfNumbers:
    MyList.append(MyNumber)
    MySet.add(MyNumber)
    if len(MySet)+NumberOfDoublesFound==len(MyList):
        print("NO")
    else:
        NumberOfDoublesFound+=1
        print("YES")    

Введите  в одну строку серез пробелы числа:12 48 46 48 78 52 48 12 98 32
NO
NO
NO
YES
NO
NO
YES
YES
NO
NO


## Урок №10. Словари

**Задание №1**

Ранее вы выполняли задание связанное с ветеринарной клиникой. В той задаче вам предстояло вывести информацию о питомце на экран. Сейчас вам необходимо создать словарь `pets = {}`

Примерный вид будет следующим:

```
pets = {
  "Имя питомца": {
    'Вид питомца': # придумайте каким образом сюда внести информацию,
    'Возраст питомца': # придумайте каким образом сюда внести информацию,
    'Имя владельца': # придумайте каким образом сюда внести информацию
  }
}
```

У вас должен получиться словарь, с ещё одним словарём внутри. То есть, есть словарь `pets`. Он в себе хранит ещё один словарь, который обозначается именем питомца. Имя питомца также нужно каким-то образом вносить туда.

Задача не будет считаться выполненной, если вы заходите сразу внести информацию, не прибегая в функции `input()`

Например: </br> 
```
pets = {
  "Мухтар": {
    "Вид питомца": "Собака",
    "Возраст питомца": 9,
    "Имя владельца": "Павел" 
  }
}
```
Так должен будет выглядеть результируюший словарь, но первоначальный его вид - пустой. Его необходимо заполнить пользовательским вводом через консоль с помощью функции `input()`, а не вписать значения уже в самом коде.

Возраст питомца должен быть типа `int`
Всё остальное - строки

Так как возраст питомца указывается типом `int`. Необходимо, в соответствии с указанным возрастом выводит `год`, `года` или `лет`. Например:

`Его возраст: 24 года` </br>
`Его возраст: 21 год` </br>
`Его возраст: 19 лет` </br>

И теперь осталось только получить всю информацию о питомце в виде строки, как из задания по `Урок №3. Ввод-вывод и базовые переменные. Задание №1`, но с небольшими изменениями. Для получения информации необходимо воспользоваться методами словаря `keys()` и `values()`:  </br>
`Это желторотый питон по кличке "Каа". Возраст питомца: 19 лет. Имя владельца: Саша`

**Решение задания № 1:**

In [95]:
def AgeString(Age):
    s=str(Age)
    l=len(list(s))
    last=int((s[(l-1)]))
    prelast=int((s[(l-2)]))
    TwoLastNumbers=int(str(prelast)+str(last))
    if last==1:
            DescriptionAge='год.'
    elif 1<last<5:
        DescriptionAge='года.'
    else:
        DescriptionAge='лет.'
    if 10<TwoLastNumbers<15:
        DescriptionAge='лет.'
    return s+" "+DescriptionAge


pets={}
NameOfPet=input('Введите имя/кличку питомца, для окончания ввода нажмите Ener: ')
while len(NameOfPet)>0:
    sub_pet={}
    sub_pet['Вид питомца']=str(input('Вид питомца? ('+NameOfPet+')'))
    sub_pet['Возраст питомца']=int(input('Возраст питомца? ('+NameOfPet+')'))
    sub_pet['Имя владельца']=str(input('Имя владельца?('+NameOfPet+')'))
    pets[NameOfPet]=sub_pet
    NameOfPet=input('Введите имя/кличку питомца, для окончания ввода нажмите Ener: ')
for i in pets:
    Kind=pets[i]['Вид питомца']
    Lord=pets[i]['Имя владельца']
    Age=pets[i]['Возраст питомца']
    OutString="Это "+Kind+' по кличке "'+i+'". Возраст питомца: '+AgeString(Age)+"Имя владельца: "+Lord+"."
    print(OutString)

Введите имя/кличку питомца, для окончания ввода нажмите Ener: Каа
Вид питомца? (Каа)желторотый питон
Возраст питомца? (Каа)34
Имя владельца?(Каа)Маугли
Введите имя/кличку питомца, для окончания ввода нажмите Ener: Шарик
Вид питомца? (Шарик)пёс
Возраст питомца? (Шарик)8
Имя владельца?(Шарик)дядя Фёдор
Введите имя/кличку питомца, для окончания ввода нажмите Ener: 
Это желторотый питон по кличке "Каа". Возраст питомца: 34 года.Имя владельца: Маугли.
Это пёс по кличке "Шарик". Возраст питомца: 8 лет.Имя владельца: дядя Фёдор.


**Задание №2**

С помощью цикла создайте словарь, в котором ключи будут, например от числа 10, до -5 (включительно). А значениями этих ключей будут сами эти числа возведённые в степени равных этим числам

Например:
```
my_dict = {
  10: 10000000000,
  9: 387420489,
  # и так далее ...
  -5: -0.00032
}
```



**Решение задания № 2:**

In [100]:
my_dict={}
for i in range(10,-6,-1):
    my_dict[str(i)]=str(i**i)
print(my_dict)

{'10': '10000000000', '9': '387420489', '8': '16777216', '7': '823543', '6': '46656', '5': '3125', '4': '256', '3': '27', '2': '4', '1': '1', '0': '1', '-1': '-1.0', '-2': '0.25', '-3': '-0.037037037037037035', '-4': '0.00390625', '-5': '-0.00032'}


## Урок №11. Функции

**Задание №1**

- Создайте функцию, которая принимает в качестве параметра - натуральное целое число.
- Данная функция находит факториал полученного числа

Например, факториал числа `3` это число `6`.

- Теперь создайте список факториалов чисел от получившегося ранее факториала `6`, до `1`.

В итоге, на вход программа получает например число `3`, возвращает число `6` (факториал числа `3`) и вам нужно сделать список из факториалов числа `6` в убывающем порядке. Находим факториал числа `6` - это `720`, затем от числа `5` - это `120` и так далее вплоть до `1`

То есть, результирующий список будет выглядеть в нашем примере так: <br>
`[720, 120, 24, 6, 2, 1]`

**Решение задания № 1:**

In [106]:
def factorial(NatureNumber):
    MyOut=1
    for Number in range(NatureNumber,0,-1):
        MyOut*=Number
    return MyOut
MyList=list()
for Number in range(int(input('Введите число от которого будем вычислять факториалы по убыванию ')),0,-1):
    MyList.append(factorial(Number))
print(MyList)

Введите число от которого будем вычислять факториалы по убыванию 6
[720, 120, 24, 6, 2, 1]


**Задание №2**

В `Урок №10. Задание №1` вы создавали словарь с информацией о питомце. Теперь нам нужна "настоящая" база данных для ветеринарной клиники.

Подробный требуемый функционал будет ниже. Пока что справка:

- Создайте функцию `create`
- Создайте функцию `read`
- Создайте функцию `update`
- Создайте функцию `delete`
- Используйте словарь `pets`, который будет предоставлен ниже, либо создайте свой аналогичный

**Функция `create`:**

Данная функция будет создавать новую запись с информацией о питомце и добавлять эту информацию в наш словарь `pets`

**Функция `read`**

Данная функция будет отображать информацию о запрашиваемом питомце в виде: <br>
`Это желторотый питон по кличке "Каа". Возраст питомца: 19 лет. Имя владельца: Саша`

**Функция `update`**

Данная функция будет обновлять информацию об указанном питомце

**Функция `delete`**

Данная функция будет удалять запись о существующем питомце

-- --
Структруа результирующего словаря `pets` будет как и в `Урок №10. Задание №1`, но с небольшим видоизменением: <br>

**Словарь pets**
```python
pets = {
    1:
        {
            "Мухтар": {
                "Вид питомца": "Собака",
                "Возраст питомца": 9,
                "Имя владельца": "Павел"
            },
        },
    2:
        {
            "Каа": {
                "Вид питомца": "желторотый питон",
                "Возраст питомца": 19,
                "Имя владельца": "Саша"
            },
        },
    # и так далее
}

```

Здесь, 1 и 2 - это идентификаторы наших питомцев. Это уникальные ключи, по которым мы сможем обращаться к нашим записям в "базе данных"

Суть будущей программы будет заключаться в следующем:
- Программа будет работать с помощью цикла while с условием `command` != 'stop', то есть до тех пор, пока на предложение ввести команду, пользователь не введёт слово `stop`
- Перед взаимодействием с "базой данных" запрашивается одна из команд в качестве пользовательского ввода. Пусть это будет переменная `command`
- Функция create должна добавлять новую информацию таким образом, чтобы идентификатор увеличивался на единицу. Чтобы у вас была возможность получать последний ключ в словаре воспользуйтесь импортом модуля `collections`. В начале вашей программы пропишите строчку: `import collection`, а в функции `create` в первых строках пропишите следующий код: <br>
```python
def create():
    last = collections.deque(pets, maxlen=1)[0]
```
`last` в данном случае и будет число последнего ключа (или в нашей логике - идентификатора записи). Именно его и необходимо будет увеличивать на единицу при добавлении следующей записи.

Как вам уже известно - суть функций заключается в том, чтобы использовать один и тот же код в нескольких местах. В данной задаче вам предстоит получать информацию о питомце несколько раз. Чтобы не повторяться в коде, вам нужно создать такие функции <br>

`get_pet(ID)`:
```python
def get_pet(ID):
  # функция, с помощью которой вы получите информацию о питомце в виде словаря
  # сделайте проверку, если питомца с таким ID нету в нашей "базе данных"
  # верните в этом случае False
  # а если питомец всё же есть в "базе данных" - верните информацию о нём
  # выглядеть это может примерно так:
  return pets[ID] if ID in pets.keys() else False
```
`get_suffix(age)`:
```python
def get_suffix(age):
  # функция, с помощью которой можно получить суффикс
  # 'год', 'года', 'лет'
  # реализацию этой функции вам предстоит придумать самостоятельно
  # функция будет возвращать соответствующую строку
  return
```
`pets_list()`:
```python
def pets_list():
  # Эта функция будет создана для удобства отображения всего списка питомцев
  # Информацию по каждому питомцу можно вывести с помощью цикла for
```

Обратите внимание, если ID не существует в словаре с питомцами - будет возникать ошибка. Вам можно от неё избавиться, если правильно составить проверочное условие. Здесь не потребуется использовать такие конструкции, как
`try`, `except`, чтобы обработать возникшую ошибку

**Решение задания № 2:**

In [51]:
import collections
def create():
    sub_pet={}
    top_sub_pet={}
    NameOfPet=str(input('Кличка питомца?'))
    sub_pet['Вид питомца']=str(input('Вид питомца? ('+NameOfPet+')'))
    sub_pet['Возраст питомца']=int(input('Возраст питомца? ('+NameOfPet+')'))
    sub_pet['Имя владельца']=str(input('Имя владельца?('+NameOfPet+')'))
    top_sub_pet[NameOfPet]=sub_pet
    
    if (len(pets)==0):
        last=1
    else:
        last = collections.deque(pets, maxlen=1)[0]
        last=int(last)+1
        
    pets[str(last)]=top_sub_pet
    
def pets_list():
    print("Список питомцев,  в текущем сеансе: ")
    for PID in pets:
        OutDic={}
        OutDic=pets[PID]
        OutString="Идетификатор(ID) питомца - "+str(PID)
        for CurPet in OutDic:
            OutOutDic={}
            OutOutDic=OutDic[CurPet]
            OutString=OutString+". Кличка питомца - "+str(CurPet)+"."
        print(OutString)
        


def read():
    MyID=input('Введите идентификатор(ID) интересующего Вас питомца: ')
    
    if (get_pet(MyID)==False):
        print("Введен несуществующий идентификатор")
        return
    else:
        MyDic={}
        MyDic=get_pet(MyID)
        for PN in MyDic:
            PetName=PN
    Lord=MyDic[PN]['Имя владельца']
    Kind=MyDic[PN]['Вид питомца']
    age=MyDic[PN]['Возраст питомца']
    OutString="Это "+Kind+' по кличке "'+PN+'". Возраст питомца: '+get_suffix(age)+"Имя владельца: "+Lord+"."
        
    print(OutString)
    
def get_pet(ID):
    return pets[ID] if ID in pets.keys() else False

def get_suffix(age):
    s=str(age)
    l=len(list(s))
    last=int((s[(l-1)]))
    prelast=int((s[(l-2)]))
    TwoLastNumbers=int(str(prelast)+str(last))
    if last==1:
            DescriptionAge='год.'
    elif 1<last<5:
        DescriptionAge='года.'
    else:
        DescriptionAge='лет.'
    if 10<TwoLastNumbers<15:
        DescriptionAge='лет.'
    return s+" "+DescriptionAge

def delete():
    MyID=input('Введите идентификатор(ID) интересующего Вас питомца: ')
    
    if (get_pet(MyID)==False):
        print("Введен несуществующий идентификатор")
        return
    else:
        pets.pop(MyID)
    
  
                
def update():
    MyID=input('Введите идентификатор(ID) интересующего Вас питомца: ')
    
    if (get_pet(MyID)==False):
        print("Введен несуществующий идентификатор")
        return
    else:
        MyDic={}
        MyDic=get_pet(MyID)
        for PN in MyDic:
            PetName=PN
        MyNewDic=MyDic[PN]
        for K in MyNewDic:
            NewVal=input('Введите новое значение параметра "'+str(K)+'".Для питомца"'+str(PN)+'" , или Enter чтобы оставить старое.')
            if (len(NewVal)>0):
                MyDic[PN][K]=NewVal 

pets={}   
command=input('Введите команду (create/read/update/delete/pets_list/get_pet/stop): ')
while command != 'stop':
    if command=='create':
        sub_pet={}
        create()
    elif command=='read':
        read()
    elif command=='update':
        update()
    elif command=='delete':
        delete()
    elif command=='pets_list':
        pets_list()
    elif command=='get_pet':
        get_pet()
    command=input('Введите команду  (create/read/update/delete/pets_list/get_pet/stop): ')
                
                         

Введите команду (create/read/update/delete/pets_list/get_pet/stop): create
Кличка питомца?Каа
Вид питомца? (Каа)питон
Возраст питомца? (Каа)34
Имя владельца?(Каа)Маугли
Введите команду  (create/read/update/delete/pets_list/get_pet/stop): create
Кличка питомца?Шарик
Вид питомца? (Шарик)пёс
Возраст питомца? (Шарик)8
Имя владельца?(Шарик)дядя Фёдор
Введите команду  (create/read/update/delete/pets_list/get_pet/stop): create
Кличка питомца?Мишка
Вид питомца? (Мишка)медведь
Возраст питомца? (Мишка)45
Имя владельца?(Мишка)Маша
Введите команду  (create/read/update/delete/pets_list/get_pet/stop): pets_list
Список питомцев,  в текущем сеансе: 
Идетификатор(ID) питомца - 1. Кличка питомца - Каа.
Идетификатор(ID) питомца - 2. Кличка питомца - Шарик.
Идетификатор(ID) питомца - 3. Кличка питомца - Мишка.
Введите команду  (create/read/update/delete/pets_list/get_pet/stop): update
Введите идентификатор(ID) интересующего Вас питомца: 1
Введите новое значение параметра "Вид питомца".Для питомца"Каа" , и

## Урок №13. Двумерные списки

**Задание №1**

С помощью цикла создайте матрицу вида 10x10 (пример): <br>
```python
[[0, -2, -5, -5, 0, -4, -1, -23, -11, -20], 
[-4, -6, 16, 9, -8, 2, 10, -32, -18, -13], 
[21, 7, -6, -3, 34, -7, -23, 25, 27, 38], 
[7, 29, 38, 19, 55, 48, 46, 49, -8, -46], 
[80, -3, 76, 46, 72, -33, 14, 0, 71, -20], 
[82, 95, 23, -6, 1, 70, 50, 16, 25, 7], 
[52, 36, -25, -35, 18, -36, 93, 11, -44, 9], 
[114, 108, 33, 61, -4, 131, 34, -26, 47, -17], 
[37, 56, 135, -12, 34, 67, -17, 14, 93, -67], 
[-4, 23, 175, -44, 30, -38, 89, 159, 28, 178]]
```

И ещё одну - такой же размерности. Числа в матрице выше привидены в качестве примера (одна из генераций)

Итого у вас должно получиться сперва две матрица одинаковой размерности

И теперь вам нужно сложить эти две матрицы в третью. Формулу сложения матриц, вы можете найти в интернете, либо посетив этот [ресурс](https://studwork.org/spravochnik/matematika/matricy/slojenie-matric)

Итак. Вы должны создать две матрицы

Чтобы заполнить матрицы различными значениями - воспользуйтесь модулем `random`

Для подключения модуля `random` в код используйте строчку, которую нужно поместить в самое начало вашего скрипта
```python
import random
```

**Для справки:** <br>
[Random. Официальная документация по модулю (англ.яз)](https://docs.python.org/3/library/random.html) <br>
[Random. Информация на русском языке](https://pythonworld.ru/moduli/modul-random.html)

Для начала вы можете просто создать переменные `matrix_1` и `matrix_2`

```python
matrix_1 = [[0, -2, -1, -6, -6, 0, -9, -8, -30, -9],
[5, 12, 4, -16, -4, -9, -16, -15, 1, -26],
[13, 39, 14, 23, -4, 40, 32, 6, -8, 23],
[13, -8, 34, 49, 30, 18, 47, 11, -24, 11],
[21, 73, 71, 61, -1, 79, -34, 22, 69, 67],
[75, 25, 25, 39, 100, -12, -21, 81, -10, 87],
[81, 63, 102, 104, 53, -44, 71, -36, -36, -9],
[7, 98, 26, -3, 128, 94, 18, -26, 14, 21],
[65, 128, 80, 124, 27, -32, 73, 59, 19, 34],
[43, 111, 38, 149, 5, 112, 79, 53, 15, 92]]

matrix_2 = [[0, 4, 6, 11, 15, 6, 9, 26, 15, 21],
[-5, 4, -15, -9, -4, 2, -8, 19, -4, -1],
[-2, -39, -19, 14, 22, 5, -34, 15, 16, -9],
[-22, -52, 11, -11, -3, 16, -11, -6, -32, -2],
[-61, -47, -5, -58, 16, -13, 28, -36, -64, 2],
[-29, 23, 19, 2, -14, -87, 7, -88, 39, 7],
[-6, 18, -97, 26, -64, 0, -72, -34, -68, -92],
[-120, -117, -72, -129, -139, 16, -61, 36, -137, -29],
[-112, -83, 7, -119, -132, -129, -143, -154, -23, -34],
[32, -67, -75, -92, 15, -163, 18, 31, -162, -16]]
```

И теперь нужно придумать, как их сложить, чтобы получить `matrix_3`:
```python
matrix_3 = [[0, 2, 5, 5, 9, 6, 0, 18, -15, 12],
[0, 16, -11, -25, -8, -7, -24, 4, -3, -27],
[11, 0, -5, 37, 18, 45, -2, 21, 8, 14],
[-9, -60, 45, 38, 27, 34, 36, 5, -56, 9],
[-40, 26, 66, 3, 15, 66, -6, -14, 5, 69],
[46, 48, 44, 41, 86, -99, -14, -7, 29, 94],
[75, 81, 5, 130, -11, -44, -1, -70, -104, -101],
[-113, -19, -46, -132, -11, 110, -43, 10, -123, -8],
[-47, 45, 87, 5, -105, -161, -70, -95, -4, 0],
[75, 44, -37, 57, 20, -51, 97, 84, -147, 76]]
```

Потренируйтесь сперва в сложении матриц, а затем попробуйте с помощью модуля `random` сгенерировать матрицы подобных размеров и уже на них применить навыки сложения `matrix_1` и `matrix_2`

Задание считается выполненным, если вы напишите алгоритм, который будет уметь как складывать матрицы, так и генерировать матрица различных размерностей. Будь то матрицы `10х10` или `4х3`


**Решение задания № 1:**

In [68]:
import random

def myr():
    return random.randint(-200,200)

def pl(t):
    for i in t:
        print(i)

W=int(input('Введите ширину матрицы: '))
H=int(input('Введите высоту матрицы: '))

matrix_1=[[myr() for i in range(W)]for i in range(H)]
matrix_2=[[myr() for i in range(W)]for i in range(H)]
matrix_3=[[0 for i in range(W)]for i in range(H)]
print()
print('Первая матрица:')
pl(matrix_1)
print()
print('Вторая матрица:')
pl(matrix_2)
matrix_3[1][1]=matrix_1[1][1]+matrix_2[1][1]
for i in range(H):
    for j in range(W):
        matrix_3[i][j]=matrix_1[i][j]+matrix_2[i][j]
print()       
print('Третья матрица(сумма матриц):')
pl(matrix_3)

Введите ширину матрицы: 10
Введите высоту матрицы: 10

Первая матрица:
[-136, -22, -164, 95, 29, 42, -39, 105, -45, -64]
[134, -141, -47, -14, -112, 186, 185, 79, -184, -174]
[-27, 102, -38, 164, -163, 72, -51, -197, -97, -155]
[164, -57, -136, -55, -39, 119, 47, 43, 199, -157]
[121, 139, 137, -96, -98, -127, 165, 148, 24, -198]
[70, 92, -56, -135, -169, 110, 17, 132, 158, 136]
[-10, 16, -111, -71, -196, -107, -101, 165, -63, 133]
[142, 10, 119, 16, 115, 156, -32, -60, -99, -140]
[-164, -77, -173, 22, -197, -38, -191, -18, -187, 30]
[116, -26, 90, 167, -13, -186, -115, 181, 62, 124]

Вторая матрица:
[-157, 160, -69, 111, 19, 146, 11, 73, 46, 137]
[65, -138, -114, -49, -121, -186, 193, 149, -68, 199]
[-190, -193, 25, 134, 151, 88, -113, -87, -4, -163]
[-171, -111, -83, 143, -189, -18, 92, 44, -172, -80]
[127, -77, 25, -122, -23, -87, 82, 61, 97, 122]
[78, 63, -128, -4, -100, -165, -181, 193, 92, -38]
[-59, 156, 142, 124, 189, -161, -57, -87, 111, -196]
[122, 37, 4, -133, -21, 100, -70, 

## Урок №14. Рекурсия

**Задание №1**


Есть последовательность
```python
my_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
```

Напишите рекурсивную функцию, которая выведет все элементы от первого до последнего и в конце отобразит сообщение
`Конец списка`, если выводить больше нечего. Циклы использовать запрещено

**Решение задания № 1:**

In [122]:
def MyOutput(my_list):
    if len(my_list)>0:
        print(my_list[0])
        my_list.pop(0)
        MyOutput(my_list)
    else:
        print('Конец списка')
        
my_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
MyOutput(my_list)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
Конец списка


## Урок №15. ООП

**Задание №1**

Есть родительский класс:

In [1]:
class Transport:

    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

Создайте объект **Autobus**, который унаследует все переменные и методы родительского класса **Transport** и выведете его.

Ожидаемый результат вывода:

**Название автомобиля: Renaul Logan Скорость: 180 Пробег: 12**

**Решение:**


In [3]:
Autobus=Transport('Renaul Logan',180,12)
print('Название автомобиля: '+Autobus.name+" Скорость:"+str(Autobus.max_speed)+" Пробег: "+str(Autobus.mileage) )

Название автомобиля: Renaul Logan Скорость:180 Пробег: 12


**Задание №2**

Создайте класс **Autobus**, который наследуется от класса **Transport**. 

Дайте аргументу **Autobus.seating_capacity()** значение  по умолчанию **50**.

Используйте переопределение метода.

Используйте следующий код для родительского класса транспортного средства:

In [95]:
class Transport:
    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

    def seating_capacity(self, capacity):
        return f"Вместимость одного автобуса {self.name}  {capacity} пассажиров"

Ожидаемый результат вывода:

**Вместимость одного автобуса Renaul Logan: 50 пассажиров**



**Решение:**

In [98]:
class Autobus(Transport):
    capacity=50
    
    def seating_capacity(self):
            return f"Вместимость одного автобуса {self.name}:{self.capacity} пассажиров"
MyBus=Autobus('Renaul Logan',180,12)
print(MyBus.seating_capacity())

Вместимость одного автобуса Renaul Logan:50 пассажиров


## Урок №16. Классы и объекты

**Задание №1**

Создайте класс Касса, который хранит текущее количество денег в кассе, у него есть методы:

- top_up(X) - пополнить на X
- count_1000() - выводит сколько целых тысяч осталось в кассе
- take_away(X) - забрать X из кассы, либо выкинуть ошибку, что не достаточно денег

In [102]:
class cashbox:
    how_many=0
        
    def __init__(self,how_many):
        self.how_many=how_many
    def top_up(self,X):
        self.how_many=self.how_many+X
    def count_1000(self):
        print(self.how_many//1000)
    def take_away(self,X):
        if self.how_many<X:
            print('ошибка - недостаточно денег')
        else:
            self.how_many=self.how_many-X
        
        
MyCash=cashbox(2000)
MyCash.count_1000()
MyCash.top_up(5000)
MyCash.count_1000()
MyCash.take_away(10000)
MyCash.count_1000()
MyCash.take_away(3000)
MyCash.count_1000()

2
7
ошибка - недостаточно денег
7
4


**Задание №2**

Создайте класс Черепашка, который хранит позиции x и y черепашки, а также s - количество клеточек, на которое она перемещается за ход

у этого класс есть методы:

- go_up() - увеличивает y на s
- go_down() - уменьшает y на s
- go_left() - уменьшает x на s
- go_right() - увеличивает y на s
- evolve() - увеличивает s на 1
- degrade() - уменьшает s на 1 или выкидывает ошибку, когда s может стать ≤ 0
- count_moves(x2, y2) - возвращает минимальное количество действий, за которое черепашка сможет добраться до x2 y2 от текущей позиции

In [120]:
class turtle:
    s=1
    def __init__(self,x,y):
        self.x=x
        self.y=y
    def go_up(self):
        self.y=self.y+self.s
    def go_down(self):
        self.y=self.y-self.s
    def go_left(self):
        self.x=self.x-self.s
    def go_right(self):
        self.x=self.x+self.s
    def evolve(self):
        self.s+=1
    def degrade(self):
        if(self.s==0):
            print("Ошибка. Шаг минимален и равен 0")
        else:
            self.s-=1
    def count_moves(self,x2, y2):
        if (self.x>x2):
            MyX=self.x-x2
        else:
            MyX=x2-self.x
        if (self.y>y2):
            MyY=self.y-y2
        else:
            MyY=y2-self.y
        return MyX+MyY
    
        
        
    def all_about(self):
        print(f"Координата X {self.x} Координата Y {self.y} Шаг {self.s}")
        
        
        
MyTurtle=turtle(1,1)
MyTurtle.all_about()
print(str(MyTurtle.count_moves(0,0)))
MyTurtle.degrade()
MyTurtle.all_about()
MyTurtle.degrade()
MyTurtle.all_about()

Координата X 1 Координата Y 1 Шаг 1
2
Координата X 1 Координата Y 1 Шаг 0
Ошибка. Шаг минимален и равен 0
Координата X 1 Координата Y 1 Шаг 0


## Урок №17. О-нотация

**Задание №1** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
m = int(input())
cnt = 0
for i in range(n):
	for j in range(m):
		cnt += 1
print(cnt)

**Ответ:**O(n*m)

**Задание №2** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
ans = n * n + 100
ans -= 200
ans *= ans
if (ans < 1000000):
	ans *= 2

**Ответ:**O(с)

**Задание №3** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
m = int(input())
k = 1000
for i in range(n):
	for j in range(m):
		k += 100

for i in range(m):
	for j in range(m):
		k -= 100

**Ответ:**O(n*m+m^2)


## Урок №18-23. Делаем Игру

Для данного задания нужно использовать программу Visual Studio Code. Вся дальнейшая работа выполняется в ней.


Официальная документация по работе с VSC: https://docs.microsoft.com/ru-ru/learn/modules/python-install-vscode/

Эмодзи можно взять с сайта: https://emojipedia.org/

